retrieval/discovery/ecs.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discovery
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/defaults"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	ecs "github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	"golang.org/x/net/context"
+	"time"
+)
+
+// ECSDiscovery periodically performs ECS-SD requests. It implements
+// the TargetProvider interface.
+type ECSDiscovery struct {
+	aws         *aws.Config
+	interval    time.Duration
+	portFrom    int64
+	portTo      int64
+	clusterName string
+}
+
+// NewECSDiscovery returns a new ECSDiscovery which periodically refreshes its targets.
+func NewECSDiscovery(conf *config.ECSSDConfig) *ECSDiscovery {
+	awsConfig := defaults.Get().Config
+
+	// override the default region
+	if conf.Region != "" {
+		awsConfig.Region = &conf.Region
+	}
+	if conf.AccessKey != "" && conf.SecretKey != "" {
+		creds := credentials.NewStaticCredentials(conf.AccessKey, conf.SecretKey, "")
+		awsConfig.Credentials = creds
+	}
+
+	return &ECSDiscovery{
+		aws:         awsConfig,
+		interval:    time.Duration(conf.RefreshInterval),
+		portFrom:    int64(conf.PortFrom),
+		portTo:      int64(conf.PortTo),
+		clusterName: conf.ClusterName,
+	}
+}
+
+// Run implements the TargetProvider interface.
+func (ed *ECSDiscovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	defer close(ch)
+
+	ticker := time.NewTicker(ed.interval)
+	defer ticker.Stop()
+
+	// Get an initial set right away.
+	tg, err := ed.refresh()
+	if err != nil {
+		log.Error(err)
+	} else {
+		ch <- []*config.TargetGroup{tg}
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := ed.refresh()
+			if err != nil {
+				log.Error(err)
+			} else {
+				ch <- []*config.TargetGroup{tg}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (ed *ECSDiscovery) refresh() (*config.TargetGroup, error) {
+	tg := &config.TargetGroup{
+		Source: *ed.aws.Region,
+	}
+	ecsService := ecs.New(session.New(), ed.aws)
+	ec2s := ec2.New(session.New(), ed.aws)
+
+	listServicesResponse, err := ecsService.ListServices(&ecs.ListServicesInput{
+		Cluster:    aws.String(ed.clusterName),
```

Comment:
The idea here was that you could have different TargetProviders for different clusters with different configs, such as the interval.